dask/dataframe/shuffle.py

```diff
@@ -674,11 +674,9 @@ def set_partitions_pre(s, divisions):
 def shuffle_group_2(df, col, ignore_index):
     if not len(df):
         return {}, df
-    ind = df[col].astype(np.int64)
+    ind = df[col].astype(np.int32)
```

Comment:
Is this generally safe to do, or can we have numbers larger than int32 max?

If we need to, pandas provides a `to_numeric(arraylike, downcast="integer')` which will automatically select the smallest dtype depending on the values. I'm not sure what the performance tradeoffs will be (and it likely wouldn't work for cudf).

Reply:
Since `col` will include integer values specifying the target partition for each row, I was assuming that there should never be more than 2 billion partitions.  However, I am also in support of removing the casting altogether - It seems reasonable to pass the responsibility of esuring the appropriate type on to the function registered to `group_split_dispatch`

Reply:
Yeah I think that's a safe assumption :)

To make sure I understand, you're saying the caller can create the int32-dtype array? If so, that sounds good to me.

Reply:
> To make sure I understand, you're saying the caller can create the int32-dtype array? If so, that sounds good to me.

I am suggesting that the logic for `group_split_dispatch` should make sure the type is appropriate (and cast it if not).  I did add the necessary change to `group_split_pandas` for this PR (pandas needs to cast to int64 to use `groupsort_indexer`).  The cudf-based logic (defined in that library) does not check at the moment, but I can certainly make the necessary changes.  Note that cudf can handle all signed integer types,  so it will only need to cast if the original type is unsigned.

Reply:
Understood, thanks.